### PR TITLE
jobs: check if a job can be marked started

### PR DESCRIPTION
**Pull request by spaskob**
PR #43929 introduced flakiness in TestTruncateWhileColumnBackfill
which is caused because a successful job is trying to be
started.

Release note: none.

### CODE REVIEW

**Review by thoszhang — COMMENTED** [review]
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44018)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @pbardea, and @spaskob)

___
*[pkg/jobs/jobs.go, line 133 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44018#-Lyeglex5E3NhN-ia-Nr:-Lyeglex5E3NhN-ia-Ns:b-yfteo6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fc0cc2cb9d9c7a4e23843086849f6bb71a0b2e0/pkg/jobs/jobs.go#L133)):*
> ```Go
> 	return j.Update(ctx, func(_ *client.Txn, md JobMetadata, ju *JobUpdater) error {
> 		if md.Status != StatusPending && md.Status != StatusRunning {
> 			return errors.New(fmt.Sprintf("job with status %s cannot be marked started", md.Status))
> ```

If we return an error here (when trying to mark a successful job as started), where does it propagate?

Basically, I'm wondering whether we should return an error or just no-op and return nil.

**Review by spaskob — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44018)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @lucy-zhang, and @pbardea)

___
*[pkg/jobs/jobs.go, line 133 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44018#-Lyeglex5E3NhN-ia-Nr:-Lyek0vsB9CCmRH7MG0d:bifhlyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fc0cc2cb9d9c7a4e23843086849f6bb71a0b2e0/pkg/jobs/jobs.go#L133)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

If we return an error here (when trying to mark a successful job as started), where does it propagate?

Basically, I'm wondering whether we should return an error or just no-op and return nil.
</blockquote></details>

I think it's an error if you try to start a job that's not supposed to be started. Usually all callers handle the error.  This call does not: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/schema_changer.go#L1008
but at least the job is not marked as StatusRunning so the test succeeds.

**Review by pbardea — APPROVED** [review]
Thanks! Looks good from my perspective. It looks like all the call sites that propagate the error should be creating a new job.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44018#-:-LyeMHp88giW27HY-yVL:blrlmoa)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @lucy-zhang, and @spaskob)

___
*[pkg/jobs/jobs.go, line 133 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44018#-LyeLThF-RMIR67bt1HJ:-LyeLThF-RMIR67bt1HK:b-eeaoyw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fc0cc2cb9d9c7a4e23843086849f6bb71a0b2e0/pkg/jobs/jobs.go#L133)):*
> ```Go
> 	return j.Update(ctx, func(_ *client.Txn, md JobMetadata, ju *JobUpdater) error {
> 		if md.Status != StatusPending && md.Status != StatusRunning {
> 			return errors.New(fmt.Sprintf("job with status %s cannot be marked started", md.Status))
> ```

nit: just a heads up that this can be simplified using `errors.Errorf("job with status %s cannot be marked started", md.Status)`

**Review by spaskob — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44018)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @lucy-zhang, and @pbardea)

___
*[pkg/jobs/jobs.go, line 133 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44018#-LyeLThF-RMIR67bt1HJ:-LyjAGTJE4ejjGIFBM-X:bd5l6mr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fc0cc2cb9d9c7a4e23843086849f6bb71a0b2e0/pkg/jobs/jobs.go#L133)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

nit: just a heads up that this can be simplified using `errors.Errorf("job with status %s cannot be marked started", md.Status)`
</blockquote></details>

thanks!

**Review by thoszhang — COMMENTED** [review]
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44018)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @lucy-zhang, and @pbardea)

**Review by thoszhang — APPROVED** [review]

**Comment by spaskob** [comment]
botrs r+

**Comment by knz** [comment]
I sincerely apologize I have to preempt this because it's doomed to fail. Let me try to merge the flake patch first and then you can retry.

bors r-

**Comment by knz** [comment]
all right you can give this a shot I think
